content/docs/reference/types/reference.mdx

```diff
@@ -140,3 +140,126 @@ You can search your reference with reference selector
+          name: 'name',
+          required: true,
+        },
+        {
+          type: 'string',
+          label: 'description',
+          name: 'description',
+          required: true,
+        },
+      ],
+    },
+```
+
+**Figure: Author Collection Schema to reference by page collection**
+
+```ts
+const referenceField = {
+  label: 'Author',
+  name: 'author',
+  type: 'reference',
+  ui: {
+    optionComponent: (props: CollectionProps, _internalSys: InternalSys) => {
+      switch (props._collection) {
+        case COLLECTIONS.AUTHOR:
+          return (
+            <AuthorCollectionCustomReference
+              name={props.name}
+              description={props.description}
+            />
+          )
+
+        case COLLECTIONS.POST:
+          return <PostCollectionCustomReference title={props.title} />
+
+        default:
+          return _internalSys.path
+      }
+    },
+  },
+  collections: ['author', 'post'],
+}
+```
+
+**Figure: Extracted reference field into a module**
+
+In this example, the Author collection and Post collection are used as a reference for the Page collection. The `optionComponent`  function allows you to create a custom reference display by providing two parameters: `props` and `_internalSys`.
+
+* `props` represents the fields from the reference collection (In this case, the fields from the Author collection, which is `name`, and `description`).
+* `_internalSys` contain various useful information for you to use in the custom component such as `file name`, `path`.
+* `AuthorCollectionCustomReference` and `PostCollectionCustomReference` are the custom components that display in the reference select dropdown (see the Image below for custom reference UI).
+
+![](https://res.cloudinary.com/forestry-demo/image/upload/v1724068658/reference_field/custom-reference-field-ui_ojtrlx.png)**Figure: UI for custom reference select**
+
+> Tips: You can do console log(props) or console log(\_internalSys) in the optionComponent to find out all the provided property and used in your custom component.
+
+### Type Safe custom reference field
+
+```ts
+// You should define the types for the fields in the author collection for typesafety
+export interface AuthorProps {
+  name: string
+  description: string
+  _collection: 'author'
+}
+
+// You should define the types for the fields in the post collection for typesafety
+export interface PostProps {
+  title: string
+  excerpt: string
+  _collection: 'post'
+}
+
+// You should define the list of collection used in the reference for typesafety
+export enum COLLECTIONS {
+  AUTHOR = 'author',
+  POST = 'post',
+}
+
+// InternalSys is from tinacms where it gives a lot of useful information for user to customize their custom component
+export interface InternalSys {
+  filename: string
+  path: string
+}
+
+export type CollectionProps = AuthorProps | PostProps
+```
+
+**Figure : Recommended Implementation for Custom Reference Selection with Type Safety.**
```

Comment:
```suggestion
### Type Safe custom reference field

The example below contains the recommended Implementation for custom reference selections, including type safety.

```ts
// You should define the types for the fields in the author collection for typesafety
export interface AuthorProps {
  name: string
  description: string
  _collection: 'author'
}

// You should define the types for the fields in the post collection for typesafety
export interface PostProps {
  title: string
  excerpt: string
  _collection: 'post'
}

// You should define the list of collection used in the reference for typesafety
export enum COLLECTIONS {
  AUTHOR = 'author',
  POST = 'post',
}

// InternalSys is from tinacms where it gives a lot of useful information for user to customize their custom component
export interface InternalSys {
  filename: string
  path: string
}

export type CollectionProps = AuthorProps | PostProps
```
```

Reply:
done removed figure text